modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java

```diff
@@ -0,0 +1,137 @@
+package org.testcontainers.consul;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.github.dockerjava.api.model.Capability;
+
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.utility.DockerImageName;
+
+
+/**
+ * GenericContainer subclass for Consul specific configuration and features. The main feature is the
+ * withPropertyInConsul method, where users can specify keys and values to be pre-loaded into Consul for
+ * their specific test scenario.
+ * <p>
+ */
+public class ConsulContainer<SELF extends ConsulContainer<SELF>> extends GenericContainer<SELF> {
+
+    private static final DockerImageName DEFAULT_IMAGE_NAME = DockerImageName.parse("consul");
+    private static final String DEFAULT_TAG = "1.10.4";
+
+    private static final int CONSUL_HTTP_PORT = 8500;
+    private static final int CONSUL_GRPC_PORT = 8502;
+
+    private Map<String, String> propertiesMap = new HashMap<>();
+    private List<String> initCommands = new ArrayList<>();
+
+    private int httpPort = CONSUL_HTTP_PORT;
+    private int grpcPort = CONSUL_GRPC_PORT;
+    private String[] startConsulCmd = new String[]{"agent", "-dev", "-client", "0.0.0.0"};
+
+    /**
+     * @deprecated use {@link #ConsulContainer(DockerImageName)} instead
+     */
+    @Deprecated
+    public ConsulContainer() {
+        this(DEFAULT_IMAGE_NAME.withTag(DEFAULT_TAG));
+    }
+
+    public ConsulContainer(String dockerImageName) {
+        this(DockerImageName.parse(dockerImageName));
+    }
+
+    public ConsulContainer(final DockerImageName dockerImageName) {
+        super(dockerImageName);
+
+        dockerImageName.assertCompatibleWith(DEFAULT_IMAGE_NAME);
+
+        // Use the status leader endpoint to verify if consul is running.
+        setWaitStrategy(Wait.forHttp("/v1/status/leader").forPort(httpPort).forStatusCode(200));
+
+        withCreateContainerCmdModifier(cmd -> cmd.withCapAdd(Capability.IPC_LOCK));
+        withEnv("CONSUL_ADDR", "http://0.0.0.0:" + httpPort);
+        withCommand(startConsulCmd);
+        withExposedPorts(httpPort, grpcPort);
+    }
+
+    @Override
+    protected void containerIsStarted(InspectContainerResponse containerInfo) {
+        addProperties();
+        runInitCommands();
```

Comment:
this is using a similar approach from vault. I would suggest to remove `addProperties` and rename `runInitCommands` with `runConsulCommands` with a general purpose.

`addProperties` is pretty convenient method to expose but it will couple the container class to one specific command from consul.